### PR TITLE
Tweak - Reflective vest and Energy Katana should reflect only in correct slots

### PR DESCRIPTION
**Pull request by BIGZi0348**
## About the PR

Added new case for reflect component to take into account - positioning, because:

Equippable items, like swords, that reflect bullets and lasers should do it only in your hands.

Equippable items, like armor, that reflect bullets and lasers should do it only while you wearing them.

## Why / Balance

Reflective vest should not work as a OP riot laser shield.

Energy katana should give you a feeling like you are cool ninja deflecting bullets and stuff, slicing them in mid air. Not a lucky frying pan from PUBG.

## Technical details

Added a bool value InRightPlace and value Placement to ReflectComponent.
Placement denotes where should reflective item reflect.

Added method IsInRightPlace() that's returns true if item is in right place and false otherwise.
It's placed in OnReflectEquipped and OnReflectHandEquipped.

Added check to TryReflectProjectile and TryReflectHitscan methods for new logic.

## Media

Current work of reflective properties on Energy Katana and Reflective Vest:

https://github.com/user-attachments/assets/7fc0848b-a579-4fea-8de4-4d4fd7409b68

New work of reflective properties on Energy Katana and Reflective Vest:

https://github.com/user-attachments/assets/51ac35ea-21a7-4ca0-9d78-7ad8992d3c87

## Requirements

- [x] I have read and I am following the [Pull Request Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html). I understand that not doing so may get my pr closed at maintainer’s discretion
- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- tweak: Reflective vest now reflects lasers only while equipped.
- tweak: Energy katana now reflects lasers and projectiles only while in hands.

### CODE REVIEW

**Comment by Plykiya** [comment]
Making this custom logic for the Reflection component isn't the way to go, we already have ways to do this with Inventory slot flags

**Comment by BIGZi0348** [comment]
> Making this custom logic for the Reflection component isn't the way to go, we already have ways to do this with Inventory slot flags

That's a valid opinion, however:

SlotFlags only take into account equipment slots (duh) and the way to make reflective item work in hands only is to use SlotFlags.NONE, which is not very intuitive. Also hypothetically if you wanted for item to work in certain inventory slots and in hands - you'll fail. Also also hypothetically if you wanted for item to not work anywhere at all - you'll fail.
And these problems could be solved with new enum which inherited from SlotFlags with addition of hands flag. *

Anyway my implementation is a new enum, which is reduced for a minimum of:
- None - doesn't work anywhere
- Hands - works in hands only
- Body - works anywhere, excluding hands
And technically 4th Hands and Body (default) - works anywhere.

Flaw in my shitcode - it's doesn't take into account possible hypothetical items that should reflect while equipped in certain slots and not while in others.

Proposal to use SlotFlags is not ideal.
My implementation is also not ideal.
The * solution is kinda massive in all places, but could be ideal solution for all hypothetical future problems.

P.S. If I had more time, I would have written a shorter letter.

**Comment by BIGZi0348** [comment]
I just had some free time in my disposal and I like this version more.

**Comment by KaiserMaus** [comment]
Still in work?

**Comment by BIGZi0348** [comment]
No work currently being made due to lack of stuff needed to be done.

Jokes aside, I did some digging and most likely this PR will remain in "open" forever.

> Still in work?

**Comment by pxc1984** [comment]
We definitely need this!

File: Resources/Prototypes/Entities/Objects/Weapons/Melee/sword.yml [hunk 1]
```diff
@@ -97,6 +97,8 @@
     - Back
     - Belt
   - type: Reflect
+    slotFlags:
+      - NONE
```
Comment (slarticodefast) [inline Resources/Prototypes/Entities/Objects/Weapons/Melee/sword.yml:101]:
```suggestion
    slotFlags:
    - NONE
```
yaml convention

File: Content.Shared/Weapons/Reflect/ReflectComponent.cs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 using Robust.Shared.Audio;
 using Robust.Shared.GameStates;
+using Content.Shared.Inventory;
```
Comment (slarticodefast) [inline Content.Shared/Weapons/Reflect/ReflectComponent.cs:3]: sort alphabetically

File: Content.Shared/Weapons/Reflect/ReflectComponent.cs [hunk 3]
```diff
@@ -16,6 +17,25 @@ public sealed partial class ReflectComponent : Component
     [ViewVariables(VVAccess.ReadWrite), DataField]
     public ReflectType Reflects = ReflectType.Energy | ReflectType.NonEnergy;
 
+    /// <summary>
+    /// Select in which inventory slots it will reflect.
+    /// By default, it will reflect in any inventory position, except pockets.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite), DataField]
```
Comment (slarticodefast) [inline Content.Shared/Weapons/Reflect/ReflectComponent.cs:24]:
```suggestion
    [DataField]
```
Datafield already includes VV readwrite nowadays.
Same below.

File: Content.Shared/Weapons/Reflect/ReflectSystem.cs [hunk 4]
```diff
@@ -239,4 +245,15 @@ private void RefreshReflectUser(EntityUid user)
...
+
+    /// <summary>
+    /// Checks if the reflective component should work in designated place.
+    /// </summary>
+    private static bool IsInRightPlace(ReflectComponent component, SlotFlags slotFlag)
+    {
+        if (slotFlag == SlotFlags.NONE)
+            return component.ReflectingInHands;
+        else
+            return (component.SlotFlags & slotFlag) == slotFlag;
```
Comment (slarticodefast) [inline Content.Shared/Weapons/Reflect/ReflectSystem.cs:257]: this means the ReflectInHands datafield will no longer work if you also set it to reflect in a slot

File: Content.Shared/Weapons/Reflect/ReflectComponent.cs [hunk 5]
```diff
@@ -16,6 +17,25 @@ public sealed partial class ReflectComponent : Component
...
+    /// Is it allowed to reflect while being in hands.
+    /// </summary>
+    [DataField, ViewVariables(VVAccess.ReadWrite), AutoNetworkedField]
+    public bool ReflectingInHands = true;
+
+    /// <summary>
+    /// Can only reflect when placed correctly.
+    /// </summary>
+    [DataField, ViewVariables(VVAccess.ReadWrite), AutoNetworkedField]
+    public bool InRightPlace = true;
```
Comment (slarticodefast) [inline Content.Shared/Weapons/Reflect/ReflectComponent.cs:37]: caching this like this is sus

**Review by slarticodefast — CHANGES_REQUESTED** [review]
In general I don't like the approach here.
The system should be changed to use an inventory relay on attempting to reflect instead of adding and removing `ReflectUserComponent` via several events. That component can then be completely removed. All we need is `ReflectComponent` to subscribe to the relay and allow the reflection when it is in the right slot defined in a single datafield.
You can relay to both the hands and the inventory using `RelaySubscriptionHelpers`.

File: Content.Shared/Weapons/Reflect/ReflectComponent.cs (hunk 2 above)
Comment (BIGZi0348) [reply Content.Shared/Weapons/Reflect/ReflectComponent.cs]: done

File: Content.Shared/Weapons/Reflect/ReflectComponent.cs (hunk 3 above)
Comment (BIGZi0348) [reply Content.Shared/Weapons/Reflect/ReflectComponent.cs]: done

File: Content.Shared/Weapons/Reflect/ReflectSystem.cs (hunk 4 above)
Comment (BIGZi0348) [reply Content.Shared/Weapons/Reflect/ReflectSystem.cs:257]: I agree the code written by me is somewhat hard to read and just like in real life there is no solutions - only trade off.
Here again 4 short lines could be replased with something better, but alas.

So here how its works:

Function IsInRightPlace have two entry points in OnReflectEquipped and OnReflectHandEquipped
Parameter SlotFlags slotFlag denotes the slotFlag where item is currently is.
In OnReflectHandEquipped parameter for slotFlag is SlotFlags.NONE and method IsInRightPlace will only do comparison of ReflectingInHands.
Otherwise it's OnReflectEquipped and item is in actual slot and the comparison goes for slotFlags of a component vs current slotFlag.

TL;DR: 
You are mistaken.

First video: katana reflecting in slotFlag.BELT and in hands.
id: EnergyKatana
- type: Reflect
  slotFlags:
  - BELT
  reflectingInHands: true

https://github.com/user-attachments/assets/10c52020-e4a6-49da-b975-7bf6504d7b8c

Second video: katana reflecting in slotFlag.BELT and NOT in hands.
id: EnergyKatana
- type: Reflect
  slotFlags:
  - BELT
  reflectingInHands: false
  
https://github.com/user-attachments/assets/399e4870-c9b9-41ba-949e-7f0c30231662

File: Content.Shared/Weapons/Reflect/ReflectComponent.cs (hunk 5 above)
Comment (BIGZi0348) [reply Content.Shared/Weapons/Reflect/ReflectComponent.cs:37]: I agree it's an unpopular way to implement this logic, however without any caching we need to run IsInRightPlace logic (with added way of finding out where item is) on every attempt to reflect.

You can give more constructive thought on how to do it, for now I only switched VVAccess to ReadOnly.

**Comment by BIGZi0348** [comment]
> In general I don't like the approach here. The system should be changed to use an inventory relay on attempting to reflect instead of adding and removing `ReflectUserComponent` via several events. That component can then be completely removed. All we need is `ReflectComponent` to subscribe to the relay and allow the reflection when it is in the right slot defined in a single datafield. You can relay to both the hands and the inventory using `RelaySubscriptionHelpers`.

Maybe in a future, to the long TODO list of nonessential stuff it goes.

**Comment by slarticodefast** [comment]
Ok, I made a refactor to the system in another PR (since I didn't just want to push major changes to your code).
Instead of doing a custom relay it now simply uses normal inventory/hand relay events and the exisiting helper functions for this.
That allows us to completely get rid of `ReflectUserComponent`, the `IsInRightPlace` function, `RefreshReflectUser`, and needs a few resolves less.

Since I based it off your work I made sure to credit you in the changelog. Would be great if you could have a look at that solution!
https://github.com/space-wizards/space-station-14/pull/37039

**Comment by BIGZi0348** [comment]
> Ok, I made a refactor to the system in another PR (since I didn't just want to push major changes to your code). Instead of doing a custom relay it now simply uses normal inventory/hand relay events and the exisiting helper functions for this. That allows us to completely get rid of `ReflectUserComponent`, the `IsInRightPlace` function, `RefreshReflectUser`, and needs a few resolves less.
> 
> Since I based it off your work I made sure to credit you in the changelog. Would be great if you could have a look at that solution! #37039

Sounds fair, then this PR could be closed and all further discussion will be taken there.
